rust/kernel/task.rs

```diff
@@ -182,11 +182,22 @@ impl Task {
         // is valid.
         let task = unsafe { &*(ktask as *const Task) }.into();
 
-        // SAFETY: Since the kthread creation succeeded, we know `ktask` is valid.
-        unsafe { bindings::wake_up_process(ktask) };
+        // Wakes up the thread, otherwise it won't run.
+        task.wake_up();
+
         guard.dismiss();
         Ok(task)
     }
+
+    /// Wakes up the task.
+    pub fn wake_up(&self) {
+        // SAFETY: By the type invariant, we know that `self.0.get()` is non-null and valid.
+        // And `wake_up_process` is safe to be called for any valid task, even if the task is
+        // running.
```

Comment:
Is there never code that expects suspended tasks to stay suspended until certain conditions are met?

Reply:
If you are talking about code that uses `Task` interfaces, then I think the code should handle the dependency that, i.e. not exposing `Task` or calling `wake_up` until certain conditions meet.

If you are talking about things inside `Task` or `task_struct`, then I think the answer is "never", `wake_up_process` can be called on any valid thread.

While we are at it, I think there is a broader question: for callback mechanism, there are usually two operations:

* Create and add the callback to some queue
* Notify the queue to run callback

, and if the safety depends on some conditions being satisfied between these two operations, which one should we mark as unsafe?

Reply:
> If you are talking about code that uses Task interfaces, then I think the code should handle the dependency that, i.e. not exposing Task or calling wake_up until certain conditions meet.

I expected that there would be an interface to list all tasks on the system, in which case not exposing the `Task` is not possible. I can't find such an interface, so maybe we are lucky with respect to that?

> , and if the safety depends on some conditions being satisfied between these two operations, which one should we mark as unsafe?

Notify as marking create and add as unsafe operation would require overhead to check that the notify was correct on wakeup, while marking notify as unsafe operation doesn't require any overhead.

Reply:
There are multiple ways that a thread can be waken up. E.g. spurious wakeup when waiting for an event, siganl while in interruptible sleep, etc. Generally it's wrong for a thread to make any assumptions about state when it's being woken up (it should always check it), so I wouldn't make `wake_up` unsafe.

Reply:
I see. In that case `wake_up` can indeed be safe.

Reply:
> > If you are talking about code that uses Task interfaces, then I think the code should handle the dependency that, i.e. not exposing Task or calling wake_up until certain conditions meet.
> 
> I expected that there would be an interface to list all tasks on the system, in which case not exposing the `Task` is not possible. I can't find such an interface, so maybe we are lucky with respect to that?
> 

Even if you get a random `Task` to wake up, it's still safe as long as the pointer still points to a valid `task_struct` (i.e. before the last  #`put_task_struct`. So I won't be worried about this.

> > , and if the safety depends on some conditions being satisfied between these two operations, which one should we mark as unsafe?
> 
> Notify as marking create and add as unsafe operation would require overhead to check that the notify was correct on wakeup, while marking notify as unsafe operation doesn't require any overhead.

I think there are two things we need to clarify here:

* The safety of a callback mechanism
* The logical correctness (or safety) of a callback user

`wake_up` is safe, if we are taking about the first thing, because the kernel API provides that (under the type invariants of `Task`). And if we are talking about the second thing, for example, some one write:

```rust
let foo = Ref::try_new(MaybeUninit::uninit())?;  // something uninit
let bar = foo.clone();

// initialize foo here, GOOD

let t = spawn(move || {
    // use bar
}); // implies a wake up

// or initialize foo here, BAD
```

It's really the caller's responsibility to make the above code safe.